code/datums/components/full_auto_fire.dm

```diff
@@ -0,0 +1,362 @@
+	
+	autofire_stat = AUTOFIRE_STAT_SLEEPING
+
+
+// There is a gun and there is a user wielding it. The component now waits for the mouse click.
+/datum/component/automatic_fire/proc/autofire_on(client/usercli)
+	switch(autofire_stat)
+		if(AUTOFIRE_STAT_ALERT, AUTOFIRE_STAT_FIRING)
+			return
+	autofire_stat = AUTOFIRE_STAT_ALERT
+	clicker = usercli
+	shooter = clicker.mob
+	RegisterSignal(clicker, COMSIG_CLIENT_MOUSEDOWN, .proc/on_mouse_down)
+	RegisterSignal(parent, COMSIG_ITEM_DROPPED, .proc/autofire_off)
+	RegisterSignal(shooter, COMSIG_MOB_LOGOUT, .proc/autofire_off)
+	parent.RegisterSignal(src, COMSIG_AUTOFIRE_ONMOUSEDOWN, /obj/item/weapon/gun/.proc/autofire_bypass_check)
+	parent.RegisterSignal(parent, COMSIG_AUTOFIRE_SHOT, /obj/item/weapon/gun/.proc/do_autofire)
+
+
+/datum/component/automatic_fire/proc/autofire_off(datum/source)
+	switch(autofire_stat)
+		if(AUTOFIRE_STAT_SLEEPING, AUTOFIRE_STAT_IDLE)
+			return
+		if(AUTOFIRE_STAT_FIRING)
+			stop_autofiring()
+	autofire_stat = AUTOFIRE_STAT_IDLE
+	if(!QDELETED(clicker))
+		UnregisterSignal(clicker, list(COMSIG_CLIENT_MOUSEDOWN, COMSIG_CLIENT_MOUSEUP, COMSIG_CLIENT_MOUSEDRAG))
+	mouse_status = AUTOFIRE_MOUSEUP //In regards to the component there's no click anymore to care about.
+	clicker = null
+	if(!QDELETED(shooter))
+		UnregisterSignal(shooter, COMSIG_MOB_LOGOUT)
+	UnregisterSignal(parent, COMSIG_ITEM_DROPPED)
+	shooter = null
+	parent.UnregisterSignal(parent, COMSIG_AUTOFIRE_SHOT)
+	parent.UnregisterSignal(src, COMSIG_AUTOFIRE_ONMOUSEDOWN)
+
+
+/datum/component/automatic_fire/proc/on_mouse_down(client/source, atom/target, turf/location, control, params)
+	var/list/modifiers = params2list(params) //If they're shift+clicking, for example, let's not have them accidentally shoot.
+	if(modifiers["shift"])
+		return
+	if(modifiers["ctrl"])
+		return
+	if(modifiers["middle"])
+		return
+	if(modifiers["alt"])
+		return
+
+	if(source.mob.in_throw_mode)
+		return
+	if(!isturf(source.mob.loc)) //No firing inside lockers and stuff.
+		return
+	if(get_dist(source.mob, target) < 2) //Adjacent clicking.
+		return
+
+	if(isnull(location)) //Clicking on a screen object.
+		if(target.plane != CLICKCATCHER_PLANE) //The clickcatcher is a special case. We want the click to trigger then, under it.
+			return //If we click and drag on our worn backpack, for example, we want it to open instead.
+		target = params2turf(modifiers["screen-loc"], get_turf(source.eye), source)
+		if(!target)
+			CRASH("Failed to get the turf under clickcatcher")
+	
+	if(SEND_SIGNAL(src, COMSIG_AUTOFIRE_ONMOUSEDOWN, source, target, location, control, params) & COMPONENT_AUTOFIRE_ONMOUSEDOWN_BYPASS)
+		return
+
+	source.click_intercepted = world.time //From this point onwards Click() will no longer be triggered.
+
+	switch(autofire_stat)
+		if(AUTOFIRE_STAT_SLEEPING, AUTOFIRE_STAT_IDLE)
+			CRASH("on_mouse_down() called with [autofire_stat] autofire_stat")

```

Comment:
More unnecessary two case switches

Reply:
👀

Reply:
Same here, the replacement would be two if checks:
```
if(autofire_stat & AUTOFIRE_STAT_SLEEPING|AUTOFIRE_STAT_IDLE)
	CRASH("on_mouse_down() called with [autofire_stat] autofire_stat")
if(autofire_stat & AUTOFIRE_STAT_FIRING)
	stop_autofiring()
```